projects/plugins/boost/app/assets/src/js/features/critical-css/lib/stores/critical-css-state.ts

```diff
@@ -1,232 +1,180 @@
+ * @param {Function}    onSuccess - Optional callback for handling the new state.
+ */
+function useCriticalCssAction<
+	ActionSchema extends z.ZodSchema,
+	ActionRequestData extends z.infer< ActionSchema >,
+>( action: string, schema: ActionRequestData, onSuccess?: ( state: CriticalCssState ) => void ) {
+	const responseSchema = z.object( {
+		success: z.boolean(),
+		state: CriticalCssStateSchema,
+		error: z.string().optional(),
+	} );
 
-		return $state;
+	// A bit annoying: you have to specify ALL template params when specifying any.
+	// Template params must be specified here, otherwise action request schema of z.void doesn't work.
+	return useDataSyncAction<
+		typeof CriticalCssStateSchema,
+		typeof responseSchema,
+		typeof schema,
+		z.infer< typeof schema >,
+		z.infer< typeof responseSchema >,
+		CriticalCssState
+	>( {
+		namespace: 'jetpack_boost_ds',
+		key: 'critical_css_state',
+		action_name: action,
+		schema: {
+			state: CriticalCssStateSchema,
+			action_request: schema,
+			action_response: responseSchema,
+		},
+		callbacks: {
+			onResult: ( result, _state ): CriticalCssState => {
+				if ( result.success ) {
+					if ( onSuccess ) {
+						onSuccess( result.state );
+					}
+
+					return result.state;
+				}
+
+				const message = result.error || __( 'Critical CSS action failed', 'jetpack-boost' );
+				return criticalCssErrorState( message );
+			},
+		},
 	} );
 }
 
-export const refreshCriticalCssState = async () => {
-	const state = await stateClient.endpoint.GET();
-	cssStateStore.override( state );
-	return state;
-};
-
-export const regenerateCriticalCss = async () => {
-	// Clear regeneration suggestions
-	suggestRegenerateDS.store.set( null );
-
-	// Immediately set the status to pending to disable the regenerate button
-	replaceCssState( { status: 'pending' } );
-
-	// This will clear the CSS from the database
-	// And return fresh nonce, provider and viewport data.
-	const freshState = await generateCriticalCssRequest();
-	if ( ! freshState ) {
-		// @REFACTORING - Handle error. Currently this dies silently.
-		return false;
-	}
+/**
+ * Hook which creates a callable action for writing generated CSS for a provider key. Returns a new
+ * async function that can be called directly.
+ */
+export function useSetProviderCssAction() {
+	return useCriticalCssAction(
+		'set-provider-css',
+		z.object( {
+			key: z.string(),
+			css: z.string(),
+		} )
+	);
+}
 
-	// We received a fresh state from the server,
-	// it's already saved there,
-	// This will update the store without triggering a save back to the server.
-	cssStateStore.override( freshState );
+/**
+ * Hook which creates a callable action for dismissing or undismissing a specific provider error.
+ */
+export function useSetProviderErrorDismissedAction() {
+	return useCriticalCssAction(
+		'set-provider-error-dismissed',
+		z.object( {
+			key: z.string(),
+			dismissed: z.boolean(),
+		} )
+	);
+}
```

Comment:
I guess this is just personal preference. But, I feel like an API like following would feel more intuitive to me.

```
export function useCriticalCssActions() {
	return {
		setProviderError: useCriticalCssAction(
			'set-provider-error-dismissed',
			z.object( {
				key: z.string(),
				dismissed: z.boolean(),
			} )
		),
		setProviderCss: useCriticalCssAction(
			'set-provider-css',
			z.object( {
				key: z.string(),
				css: z.string(),
			} )
		),
		...
	}
}
```

Then the API would look like:

```
const { setProviderError, setProviderCss } =  useCriticalCssActions();
```

Reply:
🤷‍♂️ I'll disagree this time - I prefer explicit functions that don't carry more than they need to.

Reply:
🤷‍♂️ I'll disagree this time - I prefer explicit functions that don't carry more than they need to.

Reply:
Yeah, I can see your point @haqadn - that is nice. But ultimately I don't want to create actions the caller won't use, so I think I'll have to side with Nauris on this one.